src/odemis/acq/millmng.py

```diff
@@ -0,0 +1,512 @@
+        if len(self.sites) == 0:
+            logging.debug("Site location is not provided by the user, cancelling the milling")
+            self._future._task_state = CANCELLED
+            raise CancelledError()
+
+        remaining_sites = (len(self.sites) - sites_done)
+
+        if actual_time_per_site:
+            milling_time = actual_time_per_site * remaining_sites
+        else:
+            milling_time = self._milling_time_estimate * remaining_sites
+
+        self._remaining_stage_time -= self.estimate_stage_movement_time(self.sites[sites_done])
+        total_duration = milling_time + self._remaining_stage_time
+
+        return total_duration
+
+    def estimate_stage_movement_time(self, site: CryoFeature, stage_pos_ref: tuple = None) -> float:
+        """
+        Estimation the time taken by the stage to move from current position to the location of given site
+        :param site: (CryoFeature) consists X,Y,Z coordinates for stage location in m.
+        :param stage_pos_ref: (tuple) reference position of the stage in m from where the stage moves.
+        :return: (float) time to move the stage between two points in s.
+        """
+        # current position from x,y,z of stage position and eliminating rx,ry,rz
+        if stage_pos_ref is None:
+            stage_pos = self._stage.position.value
+            current_pos = [stage_pos[an] for an in ("x", "y", "z")]
+        else:
+            current_pos = stage_pos_ref
+        target_pos = site.pos.value  # list
+        diff = [abs(target - current) for target, current in zip(target_pos, current_pos)]
+        stage_time = math.sqrt(sum(d ** 2 for d in diff)) / self._move_speed
+
+        return stage_time
+
+    def estimate_drift_time(self, setting: MillingSettings) -> float:
+        """
+        Estimate the time taken to acquire drift correction images
+        :param setting: (MillingSettings) milling settings
+        return (float): estimated time to acquire 1 anchor area
+        """
+        if setting.dcRoi.value == UNDEFINED_ROI:
+            return 0
+
+        nb_anchor_scanning = math.ceil(setting.duration.value / setting.dcPeriod.value)
+        drift_estimation = AnchoredEstimator(self._scanner, self._detector,
+                                             setting.dcRoi.value, setting.dcDwellTime.value, ANCHOR_MAX_PIXELS,
+                                             follow_drift=False)
+        return drift_estimation.estimateAcquisitionTime() * nb_anchor_scanning
+
+    def _move_to_site(self, site: CryoFeature):
+        """
+        Move the stage to the given site.
+        :param site: (CryoFeature) The site to move to.
+        :raises MoveError: if the stage failed to move to the given site.
+        """
+        target_pos = {"x": site.pos.value[0],
+                      "y": site.pos.value[1],
+                      "z": site.pos.value[2]}
+        # TODO check if it is wise to move in Z direction without affecting the OPTICAL-FIB alignment
```

Comment:
Yes, that's the point of moving in Z: to keep the optical-FIB alignment.
=> Drop this comment.